### PR TITLE
Switch oc receiver to use a worker pool instead of pure go-routines

### PR DESCRIPTION
**Pull request by sjkaris**
As discussed on #395 , use a worker pool to limit the # of go-routines.

### CODE REVIEW

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 1]
```diff
@@ -73,6 +80,20 @@ func (oci *Receiver) Export(tes agenttracepb.TraceService_ExportServer) error {
 		return err
 	}
 
+	messageChan := make(chan *data.TraceData, 64)
```
Comment (odeke-em) [inline receiver/opencensusreceiver/octrace/opencensus.go]: How did you come to choosing this capacity? A value that you gave you some optimal throughput?

Comment (odeke-em) [reply receiver/opencensusreceiver/octrace/opencensus.go]: Perhaps let's add a comment or make this a const for easier discoverability and document it too.

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 2]
```diff
@@ -93,7 +114,13 @@ func (oci *Receiver) Export(tes agenttracepb.TraceService_ExportServer) error {
 			resource = recv.Resource
 		}
 
-		go oci.export(ctxWithReceiverName, tes, lastNonNilNode, resource, recv.Spans)
```
Comment (odeke-em) [inline receiver/opencensusreceiver/octrace/opencensus.go]: Nice, thank you!

File: receiver/opencensusreceiver/octrace/opencensus.go (hunk 1 above)
Comment (pjanotti) [inline receiver/opencensusreceiver/octrace/opencensus.go]: It feels like the setup of the workers should happen before the call to Export. Or is the intent really to have a channel (and respective workers) per call to Export?

Comment (sjkaris) [reply receiver/opencensusreceiver/octrace/opencensus.go]: I went back and forth on this -- the previous code had does the grouping per connection (per export).. Thinking more on it, I think it is probably better to have the worker pool per receiver as you are suggesting.

**Comment by bogdandrutu** [comment]
Please ignore my approvals, I was testing codeowners.

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 3]
```diff
@@ -24,26 +24,53 @@ import (
...
-		opt.WithReceiver(oci)
+		opt(oci)
+	}
+
+	// Setup and startup worker pool
+	workers := make([]*receiverWorker, 0, oci.numWorkers)
+	for index := 0; index < oci.numWorkers; index++ {
+		worker := newReceiverWorker(oci)
+		go worker.listenOn(messageChan)
+		workers = append(workers)
```
Comment (pjanotti) [inline receiver/opencensusreceiver/octrace/opencensus.go]: this doesn't look correct

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 4]
```diff
@@ -107,17 +143,45 @@ func (oci *Receiver) Export(tes agenttracepb.TraceService_ExportServer) error {
...
-	tes agenttracepb.TraceService_ExportServer,
-	node *commonpb.Node,
-	resource *resourcepb.Resource,
-	spans []*tracepb.Span,
-) {
-	tracedata := data.TraceData{Node: node, Resource: resource, Spans: spans}
-	// We MUST unconditionally record metrics from this reception.
-	spansMetricsFn := internal.NewReceivedSpansRecorderStreaming(tes.Context(), receiverName)
-	spansMetricsFn(tracedata.Node, tracedata.Spans)
+// Stop stops this receivers workers
```
Comment (pjanotti) [inline receiver/opencensusreceiver/octrace/opencensus.go]: nit: suggestion: // Stop the receiver and its workers.

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 5]
```diff
@@ -24,26 +24,53 @@ import (
...
 
 // New creates a new opencensus.Receiver reference.
 func New(sr receiver.TraceReceiverSink, opts ...Option) (*Receiver, error) {
 	if sr == nil {
 		return nil, errors.New("needs a non-nil receiver.TraceReceiverSink")
 	}
-	oci := &Receiver{spanSink: sr}
+
+	messageChan := make(chan *traceDataWithCtx, 64)
+	oci := &Receiver{
```
Comment (pjanotti) [inline receiver/opencensusreceiver/octrace/opencensus.go]: nit: not from your change but `oci` was when "interceptor" was used, but we changed it to "receiver", seems a good opportunity to change it.

File: receiver/opencensusreceiver/octrace/opencensus.go [hunk 6]
```diff
@@ -107,17 +143,45 @@ func (oci *Receiver) Export(tes agenttracepb.TraceService_ExportServer) error {
...
+}
+
+func newReceiverWorker(receiver *Receiver) *receiverWorker {
+	return &receiverWorker{
+		receiver: receiver,
+		cancel:   make(chan struct{}),
+	}
+}
+
+func (rw *receiverWorker) listenOn(cn chan *traceDataWithCtx) {
```
Comment (pjanotti) [inline receiver/opencensusreceiver/octrace/opencensus.go]: nit: declare the direction of the channel on the function

**Pull request merged**